### DISCUSSION TITLE
Relationship between Vulkan API and vulkanalia

### DISCUSSION THREAD

**Discussion by sajvn0x** [KyleMayes/vulkanalia#193] [General]
what I am thinking about binding libraries, are like header files for c/c++, Is this correct? and I checked the code some files are generated using generator how generator work can you give me a brief explanation about this?

**Comment by KyleMayes** [comment]
This section of the tutorial covers the relationship between the Vulkan API and this crate in a significant amount of detail:
https://kylemayes.github.io/vulkanalia/overview.html#api-concepts

A quick summary:

- Most of the code in this crate is generated from [an XML file containing the Vulkan API definition](https://github.com/KhronosGroup/Vulkan-Docs/blob/main/xml/vk.xml)
- The official Vulkan C/C++ headers are themselves generated from that XML file, this crate has nothing to do with those headers
- `vulkanalia-sys` is pretty close to the bare minimum you need to interact with Vulkan from Rust, it just contains the necessary types (plus some trait implementations for those types like `Debug` or `Default`) and leaves everything else to the user
- `vulkanalia` is a wrapper around the raw, low-level API offered by `vulkanalia-sys` that adds very little overhead but makes Vulkan significantly easier and more idiomatic to use from Rust (plus a few convenience features like easy integration with windowing libraries like `winit`)
